test/lib/pwa.js

```diff
@@ -157,31 +158,56 @@ describe('lib.pwa', () => {
     });
   });
 
+  describe('#fetchManifest', () => {
+    afterEach(() => {
+      simpleMock.restore();
+    });
+    it('Fetches manifest directly from MANIFEST_URL', () => {
+      // Mock libManifest
+      simpleMock.mock(libManifest, 'fetchManifest').resolveWith(manifest);
+      return libPwa.fetchManifest(pwa).should.be.fulfilled.then(fetchedManifest => {
+        assert.equal(fetchedManifest, manifest);
+        assert.equal(libManifest.fetchManifest.callCount, 1);
+      });
+    });
+    it('Fails directly and looks for manifest link on START_URL', () => {
+      // Mock libManifest
+      simpleMock.mock(libManifest, 'fetchManifest').rejectWith(new Error()).resolveWith(manifest);
+      simpleMock.mock(dataFetcher, 'fetchLinkRelManifestUrl').resolveWith(START_URL);
+      return libPwa.fetchManifest(new Pwa(START_URL, manifest))
+      .should.be.fulfilled.then(fetchedManifest => {
+        assert.equal(fetchedManifest, manifest);
+        assert.equal(libManifest.fetchManifest.callCount, 2);
+        assert.equal(dataFetcher.fetchLinkRelManifestUrl.callCount, 1);
+      });
+    });
+  });
+
   describe('#save (core logic)', () => {
     afterEach(() => {
       simpleMock.restore();
     });
     it('performs all the save steps', () => {
       // Mock findByManifestUrl and bd to avoid making real calls
```

Comment:
@sebastianbenz great feedback, we keep postponing refactoring this. I like your pipeline base approach, I will take a look into it :)